scripts/backup/backup

```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+UMBREL_ROOT="$(readlink -f $(dirname "${BASH_SOURCE[0]}")/../..)"
+BACKUP_FOLDER="backup"
+BACKUP_ROOT="${UMBREL_ROOT}/${BACKUP_FOLDER}"
+BACKUP_FILE="${UMBREL_ROOT}/backup.tar.gz.pgp"
+
+check_dependencies () {
+  for cmd in "$@"; do
+    if ! command -v "$cmd" >/dev/null 2>&1; then
+      echo "This script requires \"${cmd}\" to be installed"
+      exit 1
+    fi
+  done
+}
+
+check_dependencies openssl tar gpg shuf curl
+
+# Deterministically derives 128 bits of cryptographically secure entropy
+derive_entropy () {
+  identifier="${1}"
+  umbrel_seed=$(cat "${UMBREL_ROOT}/db/umbrel-seed/seed") || true
+
+  # TODO: Check this correctly handles when seed file doesn't exist!
+  if [[ -z "$umbrel_seed" ]] || [[ -z "$identifier" ]]; then
+    echo "Missing derivation parameter, this is unsafe, exiting."
+    rm -f "${UMBREL_ROOT}/statuses/backup-in-progress"
+    exit 1
+  fi
+
+  # We need `sed 's/^.* //'` to trim the "(stdin)= " prefix from some versions of openssl
+  echo -n "${identifier}" | openssl dgst -sha256 -hmac "${umbrel_seed}" | sed 's/^.* //'
+}
+
+# Make sure an update is not in progres
+if [[ -f "${UMBREL_ROOT}/statuses/backup-in-progress" ]]; then
+    echo "A backup is already in progress. Exiting now."
+    exit 1
+fi
+
+echo "Creating lock..."
+touch "${UMBREL_ROOT}/statuses/backup-in-progress"
```

Comment:
That's a very nice solution.

We can abstract that into an `aquire_lock $identifier` function for https://github.com/getumbrel/umbrel/pull/188#discussion_r479739954

We need to be really careful with the lockfiles though. Combined with strict error reporting via `set -euo pipefail` it's very easy to accidentally end up in a situation where bash exits the script without cleaning up the lockfile and the script can never run again because the lockfile still exists.

I'm reluctant to change this right before we ship since this has been tested quite a bit and works well.

But I agree with your reasoning, we should refactor all of the bash scripts to make use of a lock function like this and implement proper error handling with something like `trap` that ensures lockfiles are always cleaned up.

Reply:
I'm quite sure `if` is an exception, but sure last moment changes are probably not the most brilliant idea :)